### PR TITLE
Add 1.8 Context Method Support

### PR DESCRIPTION
**Pull request by ninthclowd**
Hello, first of all thanks so much for building this library; your work has been a great help :)

### TL;DR
Statements executed via context will now return immediately if the context expires, the connection will be invalidated so the driver doesn't reuse it, and then in the background:
* the statement will be signaled to cancel
* the statement will be closed
* the connection will be closed

In addition this adds the ability to change isolation levels of the transaction when calling BeginTx.

Some of the thread safety has been simplified and/or removed, as go1.10 seems to make sure that statements and connections aren't used concurrently providing the SessionResetter is working correctly.

I also Vendor'd the module

### Interface Changes
The driver now implements the following interfaces:
* driver.DriverContext

The connection now implements the following interfaces:
* driver.ConnBeginTx
* driver.ConnPrepareContext
* driver.ExecerContext (although it is fast path disabled, so the driver will manage the statement)
* driver.QueryerContext (although it is fast path disabled, so the driver will manage the statement)
* driver.SessionResetter
* driver.Pinger

The statement now implements:
* driver.StmtExecContext
* driver.StmtQueryContext

### Package restructure
The package / file structure has been refactored to resemble some other sql.Driver packages out there like pq.  Some of the restructure is as follows:
* all statement receivers have been moved into the stmt.go and stmt_go18.go files
* all conn receivers have been moved into conn.go and conn_go18.go files
* the odbcstmt.go file has been flattened into stmt.go

If this release is merged, could we get a v1.0.0 tag on the current release, and a new vesion for this release so users can opt out if they desire to?

Again thanks for all your work on this.

-John

### CODE REVIEW

**Comment by ninthclowd** [comment]
I'm going to close this until I can find time to test more thoroughly :)

**Comment by ninthclowd** [comment]
Ok I was able to get all of the tests passing today on both windows (386) and linux.  I did have to add thread safety to the read operations on driver.Stats to allow the race test to pass (write was thread safe, just not read).

On windows tested using the following drivers:
* SQL Server 10.00.19041.01
* MySQL ODBC 8.0 ANSI Driver

**Comment by alexbrainman** [comment]
Hello  John,

Thank you for making the effort to create this PR.

Unfortunately the PR is too big for me to review. I have full time job and I also try to contribute to some open source projects. So I have to decide where my time goes.

Unfortunately I don't see how I can spend time reviewing PR with 20 commits and 144 files changed.

I briefly looked at your changes. And what I see is nice.

But I have been using this package for many years myself. Personally I don't find any need to change anything here. I suspect many package users are in the same situation.

And I also think that there are better alternatives to this package especially on non-Windows. So some users should be using different packages that are available,

Another problem I have is that I cannot test your changes.

I had Linux computer with docker installed where I could run MSSQL server on it. But that computer is dead now. So I have to find some Linux PC and have docker running on it, and make sure my setup works.

I also had Linux computer with unixODBC and FreeTDS installed on it, so I can run this package tests on Linux. That computer is dead too. So I would have to install all driver bits on Linux again.

I suppose I have no choice but to install all this to test your changes. But that will take time. And I would have to do it even before reviewing your code.

Anyway, if we agree to proceed with this work, I ask you to split this PR into smaller PRs. Preferably 1 commit (or a few) that completes preferably single task. So we can review the commit, run tests on it, submit it and start using it.

I don't want to break existing users (including myself) who don't care about these changes.

Please try not to add any new dependencies to this code. But let's discuss this when particular code is reviewed.

I also don't care if existing API of this package changes or not. This package was designed long time ago - when Go had no package versions support. I never even considered users who will use this package in a different way from just silently importing `github.com/alexbrainman/odbc`. I still don't care about those users. I don't plan to provide any backward compatibility for the package API. I will try not to change API, but if we need to change the API, it is fine with me.

Please let me know what you want to do.

Thank you.

Alex

**Comment by ninthclowd** [comment]
Alex, thanks for your reply.

> Unfortunately I don't see how I can spend time reviewing PR with 20 commits and 144 files changed.

Ya... I was kind of worried about that.  I think most of that is the vendor'ing though.  I'll try to think on how we could split it up and make it easier to digest.

> But I have been using this package for many years myself. Personally I don't find any need to change anything here. I suspect many package users are in the same situation.

This driver uses the deprecated driver.Execer and driver.Queryer interfaces.  As a result, modern go developers who use this library will get the impression they can use ExecContext and QueryContext, yet when they do, the methods don't return when the context is canceled.  

We fell victim to this, and originally ended up working around the issue in our web application.  However, the queries would still be running in the background, which could lead to denial of service.  This is why I undertook this effort, and we are currently using the context enable odbc driver version from my fork on our web servers with no issues.  I will definitely report if we find an issue. 

> Another problem I have is that I cannot test your changes.

That was a challenge for me as well.  FWIW, I did ensure that all of your tests pass on both windows and linux with the new version, but obviously someone else should validate the changes. 




While we are content using the fork from my repo, I imaging there are other users who are looking to use this on a web server.  Issue #137 seems to suggest this.  Personally I would like to get the value to them as well, but this is your repository, so you tell me what you want to do :)

Again, thanks for your work on this library.  While you may think there are better options available, I haven't found them.  I am very grateful for your contribution.

**Review by alexbrainman — COMMENTED** [review]
These are just general comments. I did not review code properly yet. Hopefully you can can re-arrange code so it is easier for me to review.

I managed to test your code on Windows and Linux. So at least this part is out of the way.

Thank you.

Alex

File: mssql_test.go [hunk 1]
```diff
@@ -110,26 +110,26 @@ func (params connParams) makeODBCConnectionString() string {
 	return c
```
[the remark this answers is not in the document]
Comment (ninthclowd) [reply mssql_test.go:110]: So this is my mistake; I did not realize you do not specify a minimum required go version in your go.mod.  

Is your intention to support every version of go and to not adopt new features?

[the remark this answers is not in the document]
Comment (ninthclowd) [reply mssql_test.go:110]: I would recommend you put a required version in your go.mod so users know what version is required for your module.  See https://blog.golang.org/using-go-modules

File: go.mod [hunk 2]
```diff
@@ -2,6 +2,7 @@ module github.com/alexbrainman/odbc
 
 require (
 	github.com/go-ole/go-ole v1.2.5
+	go.uber.org/atomic v1.7.0
```
[the remark this answers is not in the document]
Comment (ninthclowd) [reply go.mod:5]: sure

File: mssql_test.go [hunk 3]
```diff
@@ -1917,3 +1918,71 @@ IF @@ROWCOUNT = 0
...
 }
+
+func TestMSSQLQueryContextTimeout(t *testing.T) {
+	db, sc, err := mssqlConnect()
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer closeDB(t, db, sc, sc)
+
+	testingIssue5 = false
```
[the remark this answers is not in the document]
Comment (ninthclowd) [reply mssql_test.go]: to be explicit that that is not part of what I was testing.

File: .gitignore [hunk 4]
```diff
@@ -0,0 +1 @@
+.idea
```
[the remark this answers is not in the document]
Comment (ninthclowd) [reply .gitignore:1]: Sorry I'm used to working on jetbrains project and the .idea folder it generates in a project is localized.  I can remove this.

File: mssql_test.go [hunk 5]
```diff
@@ -1917,3 +1918,71 @@ IF @@ROWCOUNT = 0
...
+}
+
+func TestMSSQLExecContextTimeout(t *testing.T) {
+	db, sc, err := mssqlConnect()
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer closeDB(t, db, sc, sc)
+
+	testingIssue5 = false
```
[the remark this answers is not in the document]
Comment (ninthclowd) [reply mssql_test.go]: to be explicit that that is not part of what I was testing.

File: mssql_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (alexbrainman) [reply mssql_test.go:110]: I would like to support users with older versions of Go, if possible. We should support versions of Go that did not have modules, if possible. If you remove code that refers to `errors.Is`, I expect your code will build with fairly old versions of Go. That is all I am asking.

File: mssql_go18_test.go [hunk 6]
```diff
@@ -0,0 +1,220 @@
+package odbc
```
[the remark this answers is not in the document]
Comment (alexbrainman) [reply mssql_go18_test.go:1]: _go18 in a file name is pointless because it does not work - build fails. Even if you could make _go18 work, it is pointless, because I don't want to support go1.7 users. go1.7 is too old.

So, please, leave all tests in mssql_test.go.

File: .gitignore (hunk 4 above)
[the remark this answers is not in the document]
Comment (alexbrainman) [reply .gitignore:1]: I understand you have your own setup. But other people are different. Not everyone uses jetbrains. So remove it.

File: mssql_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (alexbrainman) [reply mssql_test.go]: Setting testingIssue5 to false is confusing here, because testingIssue5 is already set to false here. So code reader would be looking for some hidden meaning.

If you want to say something here, just add whatever comment you require.

File: mssql_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (alexbrainman) [reply mssql_test.go]: See my reply above.

**Comment by alexbrainman** [comment]
@ninthclowd I will wait for you to make changes as I requested.

Let me know, if you expect something different from me.

Thank you.

Alex

**Comment by ninthclowd** [comment]
I'm going to close this request as we are now maintaining our own fork with the context cancellation functionality we need and I feel it would be a pretty large effort to integrate this with the primary trunk based on your previous comments.  If I ever find the time to re-work this I will submit a different PR.  Thanks

**Comment by alexbrainman** [comment]
> I'm going to close this request as we are now maintaining our own fork with the context cancellation functionality we need and I feel it would be a pretty large effort to integrate this with the primary trunk based on your previous comments. If I ever find the time to re-work this I will submit a different PR.

SGTM. Thanks for letting me know.

Alex
